File: Ahorn/entities/RoomWrapController.jl
module VivHelperRoomWrapController
using ..Ahorn, Maple

@mapdef Entity "VivHelper/RoomWrapController" RoomWrapController(
	x::Integer,
	y::Integer,
	Top::Bool=false,
	TopOffset::Number=4.0,
	TopExitSpeedAdd::Number=15.0,
	Right::Bool=false,
	RightOffset::Number=8.0,
	RightExitSpeedAdd::Number=0.0,
	Bottom::Bool=false,
	BottomOffset::Number=12.0,
	BottomExitSpeedAdd::Number=15.0,
	Left::Bool=false,
	LeftOffset::Number=8.0,
	LeftExitSpeedAdd::Number=0.0,
	setByCamera::Bool=false,
	allEntities::Bool=false,
	ZFlagsData::String="",
	AutomateCameraTriggers::Bool=false,
	legacy::Bool=true
)

const placements = Ahorn.PlacementDict(
    "Room Wrap Controller (Viv's Helper)" => Ahorn.EntityPlacement(
        RoomWrapController
	) )

function Ahorn.selection(entity::RoomWrapController)
	x, y = Ahorn.position(entity)

	return Ahorn.Rectangle(x - 16, y - 16, 32, 32)

end

function Ahorn.renderAbs(ctx::Ahorn.Cairo.CairoContext, entity::RoomWrapController, room::Maple.Room)
	x, y = Ahorn.position(entity)
	Ahorn.drawSprite(ctx, "ahorn/VivHelper/wrap", x, y)
	# Creates rectangle showing the wrapping bounds of the level
	xW, yH = room.size
	xL = Int(get(entity.data, "LeftOffset", 8.0));
	xR = xW - Int(get(entity.data, "RightOffset", 8.0));
	yT = Int(get(entity.data, "TopOffset", 4.0));
	yB = yH - Int(get(entity.data, "BottomOffset", 12.0));
	n = []
	if(get(entity.data, "Left", false))
		push!(n, [(xL+8,yT+8), (xL+8,yB-8)])
	end
	if(get(entity.data, "Top", false))
		push!(n, [(xL+8,yT+8), (xR-8,yT+8)])
	end
	if(get(entity.data, "Right", false))
		push!(n, [(xR-8,yT+8), (xR-8,yB-8)])
	end
	if(get(entity.data, "Bottom", false))
		push!(n, [(xL+8,yB-8), (xR-8,yB-8)])
	end
	for N in n
		Ahorn.drawLines(ctx, N, ((32,178,170,127) ./ 255); thickness=2)
	end
end

end

File: Loenn/entities/RoomWrapController.lua
local drawableRectangle = require('structs.drawable_rectangle')
local drawableSprite = require('structs.drawable_sprite')
local utils = require('utils')
return {
    name = "VivHelper/RoomWrapController",
    fieldOrder = {
        "Top","Right","Bottom","Left",
        "TopOffset","TopExitSpeedAdd",
        "RightOffset","RightExitSpeedAdd",
        "BottomOffset","BottomExitSpeedAdd",
        "LeftOffset","LeftExitSpeedAdd",
        "setByCamera","allEntities","legacy",
        "ZFlagsData",
    },
    depth = -math.huge + 10,
    placements = {{
        name = "v",
        data = {
            Top = true, Right = false, Bottom = true, Left = false,
            TopOffset = 4.0, TopExitSpeedAdd = 15.0,
            RightOffset = 4.0, RightExitSpeedAdd = 15.0,
            BottomOffset = 4.0, BottomExitSpeedAdd = 15.0,
            LeftOffset = 4.0, LeftExitSpeedAdd = 15.0,
            setByCamera = false, allEntities = false, legacy = true,
            ZFlagsData = ""
        }},{
        name = "h",
        data = {
            Top = false, Right = true, Bottom = false, Left = true,
            TopOffset = 4.0, TopExitSpeedAdd = 15.0,
            RightOffset = 4.0, RightExitSpeedAdd = 15.0,
            BottomOffset = 4.0, BottomExitSpeedAdd = 15.0,
            LeftOffset = 4.0, LeftExitSpeedAdd = 15.0,
            setByCamera = false, allEntities = false, legacy = true,
            ZFlagsData = ""
        }},
        name = "b",
        data = {
            Top = true, Right = true, Bottom = true, Left = true,
            TopOffset = 4.0, TopExitSpeedAdd = 15.0,
            RightOffset = 4.0, RightExitSpeedAdd = 15.0,
            BottomOffset = 4.0, BottomExitSpeedAdd = 15.0,
            LeftOffset = 4.0, LeftExitSpeedAdd = 15.0,
            setByCamera = false, allEntities = false, legacy = true,
            ZFlagsData = ""
        }
    },
    selection = function(room,entity) return utils.rectangle(entity.x - 16, entity.y - 16, 32, 32) end,
    sprite = function(room, entity)
        local sprites = {drawableSprite.fromTexture("ahorn/VivHelper/wrap", entity)}
        local W = room.width
        local H = room.height
        local l = entity.LeftOffset
        local r = entity.RightOffset
        local t = entity.TopOffset
        local b = entity.BottomOffset
        local c = {0.67,0.7,0.25,0.75}
        if entity.Left then
            table.insert(sprites, drawableRectangle.fromRectangle("fill",0,t,l+1,H-t-b,c))
        end
        if entity.Top then
            table.insert(sprites, drawableRectangle.fromRectangle("fill",l,0,W-l-r,t+1,c))
        end
        if entity.Right then
            table.insert(sprites, drawableRectangle.fromRectangle("fill",W-r-1,t,r+1,H-t-b,c))
        end
        if entity.Bottom then
            table.insert(sprites, drawableRectangle.fromRectangle("fill",l,H-b-1,W-l-r,b+1,c))
        end
        return sprites
    end
}